File: 3525.find-x-value-of-array-ii.py
#
# @lc app=leetcode id=3525 lang=python3
#
# [3525] Find X Value of Array II
#

# @lc code=start
from typing import List

class Solution:
    def resultArray(self, nums: List[int], k: int, queries: List[List[int]]) -> List[int]:
        n = len(nums)
        a = [x % k for x in nums]

        tree = [None] * (4 * n + 10)

        def merge(left, right):
            res = [0] * (k + 1)
            res[0] = (left[0] * right[0]) % k
            for i in range(k):
                res[i + 1] = left[i + 1]
            for t in range(k):
                target = (left[0] * t) % k
                res[target + 1] += right[t + 1]
            return res

        def build(node, start, end):
            if start == end:
                p = a[start]
                tree[node] = [0] * (k + 1)
                tree[node][0] = p
                tree[node][p + 1] = 1
                return
            mid = (start + end) // 2
            build(2 * node, start, mid)
            build(2 * node + 1, mid + 1, end)
            tree[node] = merge(tree[2 * node], tree[2 * node + 1])

        def update(node, start, end, idx, val):
            if start == end:
                tree[node] = [0] * (k + 1)
                tree[node][0] = val
                tree[node][val + 1] = 1
                return
            mid = (start + end) // 2
            if idx <= mid:
                update(2 * node, start, mid, idx, val)
            else:
                update(2 * node + 1, mid + 1, end, idx, val)
            tree[node] = merge(tree[2 * node], tree[2 * node + 1])

        def query(node, start, end, l, r):
            if r < start or end < l:
                iden = [0] * (k + 1)
                iden[0] = 1
                return iden
            if l <= start and end <= r:
                return tree[node]
            mid = (start + end) // 2
            leftq = query(2 * node, start, mid, l, r)
            rightq = query(2 * node + 1, mid + 1, end, l, r)
            return merge(leftq, rightq)

        build(1, 0, n - 1)

        ans = []
        for qi in queries:
            indexi, valuei, starti, xi = qi
            new_val = valuei % k
            update(1, 0, n - 1, indexi, new_val)
            qnode = query(1, 0, n - 1, starti, n - 1)
            ans.append(qnode[xi + 1])
        return ans
# @lc code=end

File: 3525.find-x-value-of-array-ii.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 * @lc app=leetcode id=3525 lang=cpp
 *
 * [3525] Find X Value of Array II
 */

// @lc code=start
class Solution {
public:
    int K;

    struct Node {
        int prod;
        array<int, 5> cnt;
        Node(int p = 1) : prod(p) { cnt.fill(0); }
    };

    Node identity() {
        Node n;
        n.prod = (K == 1 ? 0 : 1 % K);
        n.cnt.fill(0);
        return n;
    }

    Node makeLeaf(int val) {
        Node n;
        n.prod = val % K;
        n.cnt.fill(0);
        n.cnt[n.prod] = 1; // the only non-empty prefix is itself
        return n;
    }

    Node mergeNode(const Node &a, const Node &b) {
        Node res;
        res.cnt.fill(0);
        res.prod = (int)((1LL * a.prod * b.prod) % K);

        // prefixes entirely in a
        for (int r = 0; r < K; r++) res.cnt[r] += a.cnt[r];

        // prefixes that include all of a and then a prefix of b
        for (int rb = 0; rb < K; rb++) {
            int r = (int)((1LL * a.prod * rb) % K);
            res.cnt[r] += b.cnt[rb];
        }
        return res;
    }

    vector<int> resultArray(vector<int>& nums, int k, vector<vector<int>>& queries) {
        K = k;
        int n = (int)nums.size();

        vector<Node> seg(2 * n);
        for (int i = 0; i < n; i++) seg[n + i] = makeLeaf(nums[i]);
        for (int i = n - 1; i >= 1; i--) seg[i] = mergeNode(seg[i << 1], seg[i << 1 | 1]);

        auto update = [&](int pos, int val) {
            int p = pos + n;
            seg[p] = makeLeaf(val);
            for (p >>= 1; p >= 1; p >>= 1) {
                seg[p] = mergeNode(seg[p << 1], seg[p << 1 | 1]);
                if (p == 1) break;
            }
        };

        auto query = [&](int l, int r) { // [l, r)
            Node left = identity();
            Node right = identity();
            for (l += n, r += n; l < r; l >>= 1, r >>= 1) {
                if (l & 1) left = mergeNode(left, seg[l++]);
                if (r & 1) right = mergeNode(seg[--r], right);
            }
            return mergeNode(left, right);
        };

        vector<int> ans;
        ans.reserve(queries.size());

        for (auto &q : queries) {
            int index = q[0], value = q[1], start = q[2], x = q[3];
            update(index, value);
            Node res = query(start, n);
            ans.push_back(res.cnt[x]);
        }

        return ans;
    }
};
// @lc code=end
